### DISCUSSION TITLE
Top-level function arrow syntax

### DISCUSSION THREAD

**Discussion by mfulton26** [denoland/deno#16026] [Q&A]
I see in [Deno Style Guide](https://deno.land/manual@v1.25.4/contributing/style_guide) a section called [Top-level functions should not use arrow syntax.](https://deno.land/manual@v1.25.4/contributing/style_guide#top-level-functions-should-not-use-arrow-syntax) I understand that the guide "is not meant as a general style guide for users of Deno" but when trying to make decisions on my own style guides it is often helpful to understand the reasons behind other style guides.

Why does Deno runtime and standard library use the `function` keyword at the top-level instead of arrow functions?

e.g. Is the thinking that doing so clearly marks functions as functions (readability)? Is it so that hoisting can be leveraged (some folks find such a feature desirable while others do not)?

There are scope differences at runtime but something that I was surprised to learn this morning is that top-level functions are added to global scope which I'm not sure I like (see [this answer](https://stackoverflow.com/a/38480257/3255152) to [What are the advantages/disadvantages for creating a top level function in ES6 with arrows or without?](https://stackoverflow.com/questions/38479859/what-are-the-advantages-disadvantages-for-creating-a-top-level-function-in-es6-w)).

Personally I've historically preferred `function` as I find it reads better, it is consistent with generator syntax (no arrow option there), and I can prefix a named function with `export default` (can't be done with named arrow functions).

**Comment by cjihrig** [comment]
My guess is that this rule is related to the `this` value of arrow functions. The guide says "Arrow syntax should be limited to closures." Functions often need to have a `this`, while closures generally just need access to the closed over variables.

**Reply by mfulton26** [reply]
Can you help me list good use cases for using `this` in a top-level function?

**Comment by cjihrig** [comment]
This type of thing is the first thing that comes to mind: https://github.com/denoland/deno_std/blob/6767ad853f0e0cead74281220f00aaf1efa34455/node/_fs/_fs_streams.ts#L234-L235

**Comment by nnmrts** [comment]
In my opinion there's only a few places where non-arrow-syntax functions are necessary and top-level is not one of them. The example @cjihrig linked is somewhat applicable, however this is not a top-level function, it's an object property assignment, to create a method on the `ReadStream` class prototype.

As soon as you use `this` in a top level function in ES6, you're doing something wrong in my eyes, because that's what classes are for. And if you actually want to refer to `window` or `globalThis` with your `this` in your arrow function, you should just use `window` or `globalThis` for better readability anyways.

The only places where I don't use arrow functions is where I can't: `constructor`, getter and setters, generator functions and class methods defined outside of the class body. The beauty about arrow functions is that they effectively fixed `this` in JS, because the `this` in arrow functions will always just refer to the `this` in the scope above and not sometimes to the function itself, other times to the class it is defined in, etc.

Arrow functions also follow the syntax of the rest of the language, where `x = y` means `y` is assigned to the `x` variable, while the `function` keyword somehow is able to the same without ever using the `=` symbol. For the same reason I'm writing classes in a somewhat unconvential but not less comprehensible way:

```js
const MyClass = class {
    // ...
};
```

instead of

```js
class MyClass {
    // ...
};
```
(where did the `const` go? or is this technically a `let`? could this redefine something defined with `let`? why is `MyClass` suddenly on `globalThis`? where did the `=` go???)

I can somewhat understand the one argument for `export default`:

```js
export default function myFunction() {};
```

is doable on one line and just way shorter than:

```js
const myFunction = () => {};

export default myFunction;
```

but I still write it the second way because I personally don't like defining things while exporting them at the "same time"/in the same statement, especially when there are multiple exports. Exports should be at the end of the file and "summarise" what gets exported from a file. Doing it this way is also a bit more flexible because you are able to remove and add exports without having to move the whole definition of something.

All in all I think the `function` syntax is not more readable just because it has a `function` keyword. On first glance you might think that and it's probably a bit easier for people just getting into JS from other languages, but as I said, using arrow syntax removes two problems with the "normal" way of defining functions:

1. `this` isn't some random unpredictable thing unless you understand it fully and are aware of all the edge cases.
2. The actual useful information of the statement wasn't swallowed and obfuscated, namely if it's a `const` or a `let` and if this is an assignment (`=`) to a variable name, while the somewhat useless `function` keyword **was** shortened to just `=>` - a fact that you're able to understand after 5 minutes of learning JS.

These are the things I would consider if I'd ever define a style guide like you are doing, @mfulton26 . Still opinionated of course.

**Reply by mfulton26** [reply]
Thank you for sharing your insights. Most team members I've worked with in the past (and myself too) generally agree with everything you've said which is why I was surprised to see Deno's internal style guide specify not using arrow functions on the top-level. I'm confident there are reasons but there's no current explanation as to why in the guide which prompted me to reach out and ask. (Maybe it has something to do with Rust idioms or something else specific to the stack the Deno team works with that makes it work better for them.)

**Reply by ghost** [reply]
> (Maybe it has something to do with Rust idioms or something else specific to the stack the Deno team works with that makes it work better for them.)

Well, (although a member of their team still hasn't responded) this sounds plausible: Rust generally uses functions `fn f(x, y, x) { w }` at the top-level, and closures `let f = |x, y, z| w` elsewhere.
